### PR TITLE
Bryanv/3962 server templates

### LINKED ISSUE #3962: Custom web templates with `bokeh serve`
@bryevdv and I talked about how to do custom web templates with `bokeh serve` (without having to roll a custom web server process), I'm going to try to capture the discussion here although I can't actually find the notes I thought we had.

### Overall purpose

The point of this feature is that you can make an HTML page "around" your Bokeh document, which might contain documentation, links, custom CSS, or whatever else.

We don't want to be a full-blown web framework. There will continue to be a limitation that you can't make your own arbitrary routes, and the page we let you create isn't an arbitrary template with arbitrary inputs, it's only a container for Bokeh models.

### Syntax for developers

Developers can customize their template by dropping a `main.html` into the app directory. `main.html` might look like this in the stocks demo:
```
<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="utf-8">
        <title>{{ title }}</title>
        {{ bokeh_head }}
    </head>
    <body>

        <h1>I want a giant header</h1>
        <p>Hi this is my stocks demo with custom HTML wrapper.</p>

        {{ document }}

    </body>
</html>
```

So this is a jinja2 template but Bokeh defines the variables available inside it; `{{ title }}` with document title, `{{ document }}` to embed the whole document, and a mandatory `{{bokeh_head}}` with the stuff Bokeh needs to put in the head section.

In addition to `{{ document }}`, we should consider placing all named Component models (those that you can `get_model_by_name()` on and that can be rendered) into the jinja2 template variables. Then people can insert different plots in different places in the template.

### Static assets

An elaboration on this feature (could be a separate issue/PR though) would allow CSS and JS assets to also be in the app directory so they can be referenced from the template.

### Implementation

To get things from the app directory into Bokeh, they need to be attached to a Document, so Document would have a new property `template` (this would be a python-side-only property, similar to `theme`, rather than one synced to the client like `title`). The `DirectoryHandler` will set the template property if found, in the same way it currently sets the theme property.

We already have `file_html` etc. in embed.py. There are at least two changes needed in embed.py:
 
 * rather than making the template parameter default to `_templates/file.html`, default to `document.template` (which by default would be loaded from `_templates/file.html`)
 * support higher-level template variables `bokeh_head` and `document` - `{{bokeh_head}}` is the same as current `bokeh_css` + `bokeh_js`, while `{{document}}` is the same as current `plot_div` + `plot_script`. If both old (more granular) and new (simplified) template variables are available, it would be backward compatible and also allow people to replace only CSS or only JS if they care about such things.

This should then use the custom template for both `bokeh html` and `bokeh serve`.

If static assets are supported, they would be attached to the document (maybe just their filenames are attached) and then `server/views/static_handler.py` would need to look up the session and serve static assets found on the document, if required.
If you try to `bokeh html` an app with static assets, we'd need to either embed them in the HTML or create a directory rather than a single HTML file or throw an error, but we should not silently break, we should handle it in some way.

### PR DESCRIPTION
**Pull request by bryevdv**
- [x] issues: fixes #3962
- [x] tests added / passed
- [x] release document entry (if new feature or API change)

This PR adds initial support for overriding the default `FILE` template for Bokeh apps with a custom user-specified template.

### CODE REVIEW

**Comment by bryevdv** [comment]
@birdsarah Do you think you could add some minimal template and styling around one of the apps (maybe `weather` or `movies`) ? Then we can point to it explicitly in the docs.

**Comment by bryevdv** [comment]
@bokeh/dev @havocp I feel like there might be some conceptual fuzziness between handlers, applications, and commands. Mostly this its just about who is responsible for what, and how best to generally handle combinations that "don't make sense". We seem to be getting lots of cases like this where it makes sense for "at most one handler to do XYZ" maybe there is some pattern that would help encapsulate this better.

**Comment by bryevdv** [comment]
E.g., maybe we can find a declarative way to tag handler classes by the optional features they might provide. This could help both with checking the "at most one..." cases but also help with warnings or alternate code paths (e.g a `bokeh html` command running an app with `static` resources)

**Comment by havocp** [comment]
each handler modifies the document, right? they are allowed to overwrite each other. I think the conceptually-pure approach here would be that if two handlers both set doc.template, the second one just wins. But making it an error is perhaps more useful/practical, if we believe (and this is probably true) that two templates represents a mistake.

File: examples/app/movies/static/foo.js [hunk 1]
```diff
@@ -0,0 +1,2 @@
+
+ldsfjlaskgjgldjfg
```
Comment (canavandl) [inline examples/app/movies/static/foo.js]: Was this intended to be committed?

File: bokeh/application/handlers/handler.py [hunk 2]
```diff
@@ -20,6 +21,13 @@ def static_path(self):
         else:
             return self._static
 
+    def template(self):
+        """Returns an app-specific custom Jinja template, if applicable."""
+        if self.failed:
+            return None
+        else:
+            return self._template
```
Comment (havocp) [inline bokeh/application/handlers/handler.py]: Instead of this, modify_document could set `doc.template` if appropriate; it could also raise an error if doc.template is set already. Possibly the error wouldn't be quite as clear, since we wouldn't know inside modify_document which earlier handler had already set the template.

**Comment by bryevdv** [comment]
Sure, that approach occurred to me to, it's certainly consistent with e.g., general python namespace behavior. But even if we go that route, practically speaking I think we need to at least warn if it happens.

File: examples/app/movies/static/foo.js (hunk 1 above)
Comment (bryevdv) [inline examples/app/movies/static/foo.js]: hah no it was not

File: examples/app/weather/templates/index.html [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+        <title>{{ title if title else "Bokeh Plot" }}</title>
+        {{ bokeh_css }}
+        {{ bokeh_js }}
+    </head>
+    <body>
+        <div>
+            HERE IS SOME CUSTOM STUFF
+        </div>
+        {{ plot_div|indent(8) }}
+        {{ plot_script|indent(8) }}
```
Comment (havocp) [inline examples/app/weather/templates/index.html]: to me this is too obscure, I'd love to see it simplified down to `{{bokeh_head}}`, `{{document}}`, and `{{title}}` perhaps (or other more obviously-meaningful things). I believe it's possible to do that without breaking the ability to also do the above more detailed breakdown. But it could be its own PR for sure.

Comment (havocp) [inline examples/app/weather/templates/index.html]: other than clarity, making this less granular would mean fewer apps break if the implementation details ever change...

Comment (bryevdv) [inline examples/app/weather/templates/index.html]: Yes sorry I should have specified, I am definitely wanting to spread things out over a few independent PRs. This one make a minimal change to add a pretty important capability, do intend to refine it more.

File: bokeh/application/handlers/handler.py (hunk 2 above)
Comment (bryevdv) [inline bokeh/application/handlers/handler.py]: Could add debug logs for any template found, and a warning if it is being overwritten?

File: examples/app/movies/static/foo.js (hunk 1 above)
Comment (havocp) [inline examples/app/movies/static/foo.js]: this code is genius though

**Comment by havocp** [comment]
A kind of weird thing is that people can change curdoc().template in a callback, which would only ever do anything if the same session was reloaded, so it would probably be pointless. Not sure we could/should bother to do anything about it.

File: bokeh/document.py [hunk 4]
```diff
@@ -356,6 +359,17 @@ def title(self, title):
...
     @property
+    def template(self):
+        return self._template
+
+    @template.setter
+    def template(self, template):
+        if not isinstance(template, jinja2.Template):
+            raise ValueError("Document templates must be Jinja2 Templates")
+        self._template = template
+        # TODO (bev) template change eventing?
```
Comment (havocp) [inline bokeh/document.py]: I can't think of where change notification would be practically useful right now, fwiw

**Comment by havocp** [comment]
should `bokeh html` use the template too?

**Comment by bryevdv** [comment]
Well so that's my question, is a template a property of the `Document`?  or of an `Application`? Could make `template` private, only updatable by handlers using private APIs

**Comment by bryevdv** [comment]
Yes `bokeh html` should use template for sure

**Comment by havocp** [comment]
Either way can probably be made to work... if the static resources are on application and not doc, maybe it makes sense to keep it with those. Document could try to be more only the thing that's modified on the fly and synced over the websocket, Application would provide the HTML "container" around the document ?
The reason i was thinking it had to be on Document was so the various output modes (server, standalone html, etc.) would just work, but as we discussed when doing the static resources PR we can probably solve that anyway when putting this stuff on Application.

**Comment by bryevdv** [comment]
> A kind of weird thing is that people can change curdoc().template in a callback, which would only ever do anything if the same session was reloaded

For sure, tho being able to set it programmatically at the end of an app script seems useful. I could imagine ways to warn about this, though...  Setting an `_initialized` flag as the last step of creating a doc in a session, and having the `template` setter check that and warning if it is changed once initialized.

**Comment by bryevdv** [comment]
@bokeh/dev I think there may eventually be tweaks to make to the implementation, but I think the user-facing interface reasonable as-is.  I've simplified some things as described above (i.e., handlers just update templates in order without error) I'd like to merge this PR as it is and make incremental improvements.

**Comment by bryevdv** [comment]
@bokeh/dev I will plan to merge this tomorrow sometime absent objection.
